mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,163 @@
+        }
+        rewriter.create<scf::YieldOp>(newForOp.getLoc(), newYieldOperands);
+        rewriter.eraseOp(clonedYieldOp);
+
+        // 6. Retensorize results after for loop
+        rewriter.setInsertionPointAfter(forOp);
+        for (std::size_t index : indices) {
+            mlir::Value for_result = newForOp->getResult(index);
+            auto from_elem_op = rewriter.create<tensor::FromElementsOp>(
+                newForOp->getLoc(), RankedTensorType::get({}, for_result.getType()), for_result);
+            rewriter.replaceUsesWithIf(
+                forOp->getResult(index), from_elem_op->getResult(0),
+                [&](OpOperand &op) { return !isa<tensor::FromElementsOp>(op.getOwner()); });
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+    }
+
+    void detensorizeIfOp(scf::IfOp if_op, IRRewriter &rewriter)
+    {
+        auto isScalarTensor = [](Value v) {
+            if (!::llvm::isa<TensorType>(v.getType())) {
+                return false;
+            }
+            return dyn_cast<TensorType>(v.getType()).getRank() == 0;
+        };
+
+        if_op->walk([&](scf::YieldOp yield_op) {
+            // Loop over results
+            std::size_t i_result = 0;
+            for (mlir::Value operand : yield_op.getOperands()) {
+                if (isScalarTensor(operand)) {
+                    auto if_result = if_op->getResult(i_result);
+                    // Detensorize operand: extract tensor element before yielding
+                    {
+                        rewriter.setInsertionPoint(yield_op);
+                        mlir::Operation *extract_op = rewriter.create<tensor::ExtractOp>(
+                            yield_op->getLoc(), operand, ValueRange{});
+                        rewriter.replaceUsesWithIf(
+                            operand, extract_op->getResult(0),
+                            [&](OpOperand &op) { return op.getOwner() == yield_op; });
+                        const mlir::Type type = extract_op->getResult(0).getType();
+                        rewriter.modifyOpInPlace(
+                            yield_op, [&]() { yield_op->getOperand(i_result).setType(type); });
+                        rewriter.modifyOpInPlace(if_op, [&]() { if_result.setType(type); });
+                    }
+                    // Retensorize operand: reconstruct tensor after the for body
+                    {
+                        rewriter.setInsertionPointAfter(if_op);
+                        auto from_elem_op = rewriter.create<tensor::FromElementsOp>(
+                            if_op->getLoc(), RankedTensorType::get({}, if_result.getType()),
+                            if_result);
+                        rewriter.replaceUsesWithIf(
+                            if_result, from_elem_op->getResult(0), [&](OpOperand &op) {
+                                return !isa<tensor::FromElementsOp>(op.getOwner());
+                            });
```

Comment:
Sorry I mean the `if` part in `replaceUsesWithIf` is not required. 

If the original of returned tensors
```
%1 = scf.if %0 -> (tensor<f64>) {
   ...
}
op.use1 %1 : tensor<f64>
op.use2 %1 : tensor<f64>
```
and now changed it to return a scalar and generate the `from_elements op` (but with the old values still around)
```
%1 = scf.if %0 -> (tensor<f64>) {
   ...
}
%s_1 = scf.if %0 -> (f64) {
   ...
}
%t_1 = tensor.from_elements %s_1 : tensor<f64>
op.use1 %1 : tensor<f64>
op.use2 %1 : tensor<f64>
```
What you want to do is go in and fix up the broken references with:
```
rewriter.replaceAllUsesWith(%1, %t_1)
```
which will only affect users of the value, so only the last two lines :
```
op.use1 %t_1 : tensor<f64>
op.use2 %t_1 : tensor<f64>
```

Reply:
Got it. This will work after refactoring, but not with the in-place implementation because you could end up with something like 
```
%t_1 = tensor.from_elements %1 : tensor<f64>
%t_2 = tensor.from_elements %t_1 : tensor<f64>
```
for example.

Reply:
Oh, that would be very strange, maybe another good reason not to do the in-place modification 😅